File: src/main/java/database/SerializeProfile_MySQL.java
package database;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.Statement;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import java.sql.SQLIntegrityConstraintViolationException;

import core.utils.Helper;
import core.utils.DoubleArraySerializer;
import core.utils.Genome;
import core.utils.Perspectives;
import core.utils.Profile;
import core.utils.Transcriptome;

/**
 * Interacts with MySql DB, by saving and loading the profile of a clone.
 * @author noemi
 *
 */
public class SerializeProfile_MySQL {
	static final CharSequence TABLE_PLACEHOLDER = "XX";
	private static final CharSequence CLONEID_PLACEHOLDER = "YY";

	private String wRITE_PROFILE_LOCI_SQL = "INSERT INTO Loci(content, hash) VALUES(?, ?)";

	private String rEAD_PROFILE_LOCI_SQL = "SELECT content FROM Loci WHERE id = ?";

	private String wRITE_PROFILE_SQL = "UPDATE XX SET profile= ?, profile_hash= ?, profile_loci=? WHERE cloneID=YY";

	private String rEAD_PROFILE_SQL = "SELECT profile,whichPerspective,profile_loci FROM XX WHERE cloneID = YY";


	static Map<Integer, String[]> s_cmap = new HashMap<Integer, String[]>();
	static Map<Integer, Integer> saved_loci_set = new HashMap<Integer, Integer>();
	static Map<Integer, Integer> saved_profile_set = new HashMap<Integer, Integer>();

	public SerializeProfile_MySQL(int cloneID, String cloneClass) {
		String tableName = CLONEID.getTableNameForClass(cloneClass);
		this.wRITE_PROFILE_SQL = this.wRITE_PROFILE_SQL.replace(TABLE_PLACEHOLDER, tableName);
		this.wRITE_PROFILE_SQL = this.wRITE_PROFILE_SQL.replace(CLONEID_PLACEHOLDER, cloneID + "");

		this.rEAD_PROFILE_SQL = this.rEAD_PROFILE_SQL.replace(TABLE_PLACEHOLDER, tableName);
		this.rEAD_PROFILE_SQL = this.rEAD_PROFILE_SQL.replace(CLONEID_PLACEHOLDER, cloneID + "");
	}

	/**
	 * Saves profile to DB
	 * @param conn
	 * @param object
	 * @throws Exception
	 */
	public void writeProfile2DB(Connection conn, Profile object) throws Exception {

		String[] loci = object.getLoci();

		int hash = Arrays.deepHashCode(loci);
		Integer SavedLociId = SerializeProfile_MySQL.saved_loci_set.get(hash);
		int lociID = -1;
		if (SavedLociId == null) {
			PreparedStatement pstmt = conn.prepareStatement(wRITE_PROFILE_LOCI_SQL, Statement.RETURN_GENERATED_KEYS);
			// Profile loci
			pstmt.setBytes(1, Helper.string2byte(loci));
			pstmt.setInt(2, hash);
			try {
				pstmt.executeUpdate();
				ResultSet rs = pstmt.getGeneratedKeys();
				rs.next();
				lociID = rs.getInt(1);
			} catch (SQLIntegrityConstraintViolationException e) {
				pstmt = conn.prepareStatement("SELECT id FROM Loci WHERE hash = " + hash);
				ResultSet rs = pstmt.executeQuery();
				rs.next();
				lociID = rs.getInt(1);
			}
			SerializeProfile_MySQL.saved_loci_set.put(hash, lociID);
			System.out.printf("ADDED LOCI:%d:%d:%d\n", lociID, String.join(":", loci).length(), hash);
			pstmt.close();
		}else{
			lociID = SavedLociId.intValue();
			// System.out.printf("SAVED LOCI:%d:%d:%d\n", lociID, 0, hash);
		}


		// Profile
		Double[] values = object.getValues();
		int hash_profile = Arrays.deepHashCode(values);
		byte[] compressedValues1 = DoubleArraySerializer.serialize(values);
		Integer SavedProfile = SerializeProfile_MySQL.saved_profile_set.get(hash_profile);
		if (SavedProfile == null) {
			PreparedStatement pstmt = conn.prepareStatement(wRITE_PROFILE_SQL);
			pstmt.setBytes(1, compressedValues1);
			pstmt.setInt(2, hash_profile);
			pstmt.setInt(3, lociID);
			pstmt.executeUpdate();
			pstmt.close();
		}
	}

	private Profile createProfile(String perspective, String[] loci, Double[] object) {
		Profile p = null;
		switch (Perspectives.valueOf(perspective)) {
			case Identity:
			case MorphologyPerspective:
				p = new Profile(loci);
				break;
			case GenomePerspective:
			case KaryotypePerspective:
			case ExomePerspective:
				p = new Genome(loci);
				break;
			case TranscriptomePerspective:
				p = new Transcriptome(loci);
				break;
			default:
				throw new IllegalArgumentException("Unknown perspective: ");
		}

		p.setValues(object);

		return p;
	}

	/**
	 * Loads profile from DB
	 *
	 * @param conn
	 * @return
	 * @throws Exception
	 */
	public Profile readProfileFromDB(Connection conn) throws Exception {
		// Read profile values
		PreparedStatement pstmt = conn.prepareStatement(rEAD_PROFILE_SQL);
		ResultSet rs = pstmt.executeQuery();
		rs.next();
		Double[] object = DoubleArraySerializer.deserialize(rs.getBytes(1));
		String perspective = rs.getString(2);
		int lociID = rs.getInt(3);
		rs.close();

		String[] loci = SerializeProfile_MySQL.s_cmap.get(lociID);

		if (loci == null) {
			// Read profile loci
			pstmt = conn.prepareStatement(rEAD_PROFILE_LOCI_SQL);
			pstmt.setInt(1, lociID);
			rs = pstmt.executeQuery();
			rs.next();
			loci = Helper.byte2String(rs.getBytes(1));
			SerializeProfile_MySQL.s_cmap.put(lociID, loci);
			pstmt.close();
		}

		return createProfile(perspective, loci, object);
	}

}

File: webportal/frontend/src/lib/cloneid/database/SerializeProfile_MySQL.js
import { Genome } from '../core/utils/Genome';
import { Helper } from '../core/utils/Helper';
import { Perspectives } from '../core/utils/Perspectives';
import { Profile } from '../core/utils/Profile';
import { Transcriptome } from '../core/utils/Transcriptome';
import { CLONEID } from './CLONEID';
import { QueryManager } from './QueryManager';

String.prototype.hashCode = function() {
    var hash = 0, i, chr;
    if (this.length === 0) return hash;
    for (i = 0; i < this.length; i++) {
        chr   = this.charCodeAt(i);
        hash  = ((hash << 5) - hash) + chr;
        hash |= 0;
    }
    return hash;
};


export class SerializeProfile_MySQL {

    static #TABLE_PLACEHOLDER = "XX";
	static #CLONEID_PLACEHOLDER = "YY";

	#wRITE_PROFILE_LOCI_SQL = "INSERT INTO Loci(content, hash) VALUES(?, ?)";
	#rEAD_PROFILE_LOCI_SQL = "SELECT content FROM Loci WHERE id = ?";
	#wRITE_PROFILE_SQL = "UPDATE XX SET profile= ?, profile_hash= ?, profile_loci=? WHERE cloneID=YY";
	#rEAD_PROFILE_SQL = "SELECT profile,whichPerspective,profile_loci FROM XX WHERE cloneID = YY";
	#clone = 0;

    constructor(cloneID, cloneClass) {
        this.wRITE_PROFILE_LOCI_SQL = "INSERT INTO Loci(content, hash) VALUES(?, ?)";
        this.rEAD_PROFILE_LOCI_SQL = "SELECT content FROM Loci WHERE id=?";
        this.wRITE_PROFILE_SQL = "UPDATE XX SET profile=?, profile_hash=?, profile_loci=? WHERE cloneID=YY";
        this.rEAD_PROFILE_SQL = "SELECT profile,whichPerspective,profile_loci FROM XX WHERE cloneID=YY";

        this.clone = cloneID;
        const tableName = CLONEID.getTableNameForClass(cloneClass);

		this.wRITE_PROFILE_SQL = this.wRITE_PROFILE_SQL.replace(SerializeProfile_MySQL.#TABLE_PLACEHOLDER, tableName);
		this.wRITE_PROFILE_SQL = this.wRITE_PROFILE_SQL.replace(SerializeProfile_MySQL.#CLONEID_PLACEHOLDER, cloneID);

		this.rEAD_PROFILE_SQL = this.rEAD_PROFILE_SQL.replace(SerializeProfile_MySQL.#TABLE_PLACEHOLDER, tableName);
		this.rEAD_PROFILE_SQL = this.rEAD_PROFILE_SQL.replace(SerializeProfile_MySQL.#CLONEID_PLACEHOLDER, cloneID+"");
    }

    /**
     * Saves profile to DB
     * @param {Connection} conn
     * @param {Profile} object
     * @throws {Exception}
     */
    async writeProfile2DB(conn, object) {
        const hash = JSON.stringify(object.getLoci()).hashCode();
        let pstmt = await conn.prepareStatement(this.wRITE_PROFILE_LOCI_SQL, Statement.RETURN_GENERATED_KEYS);
        
        const lociBytes = Helper.string2byte(object.getLoci());
        pstmt.setBytes(1, lociBytes);
        pstmt.setInt(2, hash);
        let lociID = -1;
        
        try {
            await pstmt.executeUpdate();
            const rs = await pstmt.getGeneratedKeys();
            rs.next();
            lociID = rs.getInt(1);
        } catch
        {
            pstmt = await conn.prepareStatement("SELECT id FROM Loci WHERE hash = " + hash);
            const rs = await pstmt.executeQuery();
            rs.next();
            lociID = rs.getInt(1);
        }

        pstmt = await conn.prepareStatement(this.wRITE_PROFILE_SQL);
        pstmt.setInt(2, JSON.stringify(object.getValues()).hashCode());
        pstmt.setInt(3, lociID);
        pstmt.setBytes(1, Helper.double2byte(object.getValues()));
        await pstmt.executeUpdate();

        pstmt.close();
    }


    /**
     * Loads profile from DB
     * @param {Connection} conn
     * @returns {Profile}
     * @throws {Exception}
     */
    async readProfileFromDB(conn) {
        let stmt = this.rEAD_PROFILE_SQL;
        let rs = await QueryManager.executeQuery(stmt);
        rs.next();
        let object = Helper.byte2double(rs.getBytes(1));
        let perspective = rs.getString(2);
        let lociID = rs.getInt(3);

        stmt = this.rEAD_PROFILE_LOCI_SQL;
        stmt = stmt.replace('?', lociID.toString());
        rs = await QueryManager.executeQuery(stmt);
        rs.next();
        let loci = Helper.byte2String(rs.getBytes(1));
        let p = null;
        if (perspective === Perspectives.Identity) {
            p = new Profile(loci);
        } else if (
            perspective === Perspectives.GenomePerspective ||
            perspective === Perspectives.KaryotypePerspective ||
            perspective === Perspectives.ExomePerspective
        ) {
            p = new Genome(loci);
        } else if (perspective === Perspectives.TranscriptomePerspective) {
            p = new Transcriptome(loci);
        } else if (perspective === Perspectives.MorphologyPerspective) {
            p = new Profile(loci);
        }
        
        for (let i = 0; i < object.length; i++) {
            p.modify(i, object[i]);
        }
        return p;
    }
}
